Firebase/Storage/FIRStorageReference.m

```diff
@@ -331,6 +333,91 @@ - (void)downloadURLWithCompletion:(FIRStorageVoidURLError)completion {
+  if (maxResults <= 0 || maxResults > 1000) {
+    [NSException raise:NSInvalidArgumentException
+                format:@"'maxResults' must be between 1 and 1000 inclusive."];
+  }
+
+  FIRStorageListTask *task =
+      [[FIRStorageListTask alloc] initWithReference:self
+                                     fetcherService:_storage.fetcherServiceForApp
+                                      dispatchQueue:_storage.dispatchQueue
+                                           pageSize:@(maxResults)
+                                  previousPageToken:nil
+                                         completion:completion];
+  [task enqueue];
+}
+
+- (void)listWithMaxResults:(int64_t)maxResults
+                 pageToken:(NSString *)pageToken
+                completion:(FIRStorageVoidListError)completion {
+  if (maxResults <= 0 || maxResults > 1000) {
+    [NSException raise:NSInvalidArgumentException
+                format:@"'maxResults' must be between 1 and 1000 inclusive."];
+  }
+
+  FIRStorageListTask *task =
+      [[FIRStorageListTask alloc] initWithReference:self
+                                     fetcherService:_storage.fetcherServiceForApp
+                                      dispatchQueue:_storage.dispatchQueue
+                                           pageSize:@(maxResults)
+                                  previousPageToken:pageToken
+                                         completion:completion];
+  [task enqueue];
+}
+
+- (void)listAllWithCompletion:(FIRStorageVoidListError)completion {
+  NSMutableArray *prefixes = [NSMutableArray new];
+  NSMutableArray *items = [NSMutableArray new];
+
+  __weak FIRStorageReference *weakSelf = self;
+
+  __block FIRStorageVoidListError paginatedCompletion =
+      ^(FIRStorageListResult *listResult, NSError *error) {
+        if (error) {
+          completion(nil, error);
+        }
+
+        FIRStorageReference *strongSelf = weakSelf;
+        if (!strongSelf) {
+          return;
+        }
+
+        [prefixes addObjectsFromArray:listResult.prefixes];
+        [items addObjectsFromArray:listResult.items];
+
+        if (listResult.pageToken) {
+          FIRStorageListTask *nextPage = [[FIRStorageListTask alloc]
+              initWithReference:self
+                 fetcherService:strongSelf->_storage.fetcherServiceForApp
+                  dispatchQueue:strongSelf->_storage.dispatchQueue
+                       pageSize:nil
+              previousPageToken:listResult.pageToken
+                     completion:paginatedCompletion];
+          [nextPage enqueue];
+        } else {
+          FIRStorageListResult *result = [[FIRStorageListResult alloc] initWithPrefixes:prefixes
+                                                                                  items:items
+                                                                              pageToken:nil];
+          paginatedCompletion = nil;
```

Comment:
Just wonder why assigning `nil` to `paginatedCompletion` is required?

Reply:
Looks correct to me. Just was curious about details, so I know what to watch out for if I need to implement something similar. 
Nit: I would be happy to see a comment for it, but up to you.

Reply:
Added comment.

Reply:
Thanks!